Fix flashlight clipping through world/objects while attached to gun muzzle

Ephememory opened the pull request:
This fixes a minor issue when the view flashlight clips through object/world along with the muzzle, especially in the case of long weapons like sniper rifles.

Originally, I went with the system that I ported from Source1 (in this case it would be how Left 4 Dead does it) to my gamemode. Since there is currently no way I know of to disable a light from affecting certain models I went with this system since it would otherwise clip backwards into the viewmodel and look ugly with weird lighting.

mzegar commented:
This works good, I just don't know whether or not we want to move the flashlight. I'll leave it up for @kolexxx to decide. I guess it just seems potentially odd to me to move the flashlight from the gun barrel to the eyes.

Ephememory commented:
> This works good, I just don't know whether or not we want to move the flashlight. I'll leave it up for @kolexxx to decide. I guess it just seems potentially odd to me to move the flashlight from the gun barrel to the eyes.

Yeah, it's very odd and jarring, I get ya.

 If you would prefer the other way: where the flashlight is still at the barrel on the X and Y axis, and only gets moved on the Z axis: the same way something like L4D2 does it, example:

https://user-images.githubusercontent.com/66768086/182537319-79afb991-79cc-4fcf-afb6-0bb9de3c26c6.mp4


The only issue with this is that I don't think we currently have a way to render viewmodels separately, and the viewlight will illuminate the viewmodel (sometimes from the inside, which is especially ugly.)

mzegar commented:
I wonder if we could move the view model itself backward? Yeah, it may illuminate the viewmodel, do you still want to attempt it?

Ephememory commented:
> I wonder if we could move the view model itself backward? Yeah, it probably illuminate the viewmodel, do you still want to attempt it?

Sure thing, I already wrote the code for it yesterday but scrapped it. I'll restore and then we'll see what we can do.

Ephememory commented:
https://user-images.githubusercontent.com/66768086/182556104-2092bac5-7fcc-43ba-acae-3b75eba15fd6.mp4

Alright, here is what I had the other day but better since I had missed a case where the origin is clipping through doors/walls.

mzegar commented:
This works much better imo, really neat stuff.

mzegar commented on code/Player/Player.Flashlight.cs:
```diff
@@ -71,6 +71,9 @@ protected void DeleteFlashlight()
 		_viewLight = null;
 	}
 
+	[ConVar.Client( "ttt_visualizeflashlight" )]
+	public static bool VisualizeFlashlight { get; set; } = false;
```
I think there are some `debug` preprocessors you can use for this, but, I think we'd just want to remove this.

Ephememory replied:
Yep, no worries. Just left the option there in case you wanted to include it.

mzegar reviewed (commented):
I don't think I have anything else to comment on.

mzegar commented:
@Ephememory Just a thought, what do you think about making this generic and applying it to dropped ammo & dropped equipment?

For example, right now we always spawn dropped `Ammo` X distance away from the player, we instead should do something similar and move the drop location closer to the player if they are close to a wall. I think this might be similar enough to make a generic solution for it?

We'd want to handle
1. Dropped weapons (g)
2. Dropped ammo (shift+g)
3. C4 (when you left click? all deployables inherit from `Deployable`)

What do you think?

Ephememory commented:
> @Ephememory Just a thought, what do you think about making this generic and applying it to dropped ammo & dropped equipment?
> 
> For example, right now we always spawn dropped `Ammo` X distance away from the player, we instead should do something similar and move the drop location closer to the player if they are close to a wall. I think this might be similar enough to make a generic solution for it?
> 
> We'd want to handle
> 
> 1. Dropped weapons (g)
> 2. Dropped ammo (shift+g)
> 3. C4 (when you left click? all deployables inherit from `Deployable`)
> 4. Might be a bit more extra work, but `Hands` both dropping and throwing in `Hands.GrabbableProp.cs`
> 
> What do you think?

Yeah, I think that's a good idea. I'll get to work on that.

Ephememory commented:
Alright, that was a good idea. I'll just use this PR for the other items on the list and rename it.

Ephememory commented:
Works well enough as is, actually. Guess just keep the maxOffset reasonably low enough to where the maximum ends up near the center of the player. Dunno if that means leaving up the physics engine's resolver up to chance.

https://user-images.githubusercontent.com/66768086/182839415-3f1d5a71-3973-47e7-9bd5-33745b46d94b.mp4

Ephememory commented:
Reset branch to https://github.com/sbox-TTT/TTT/pull/148/commits/8e569e13e47037870c031d675d32da58aafa6f69

kolexxx commented on code/Player/Player.Flashlight.cs:
```diff
@@ -82,11 +82,58 @@ private void Frame()
+		// If there IS something between our eyes and the muzzle, add the distance.
+		if ( muzzleTrace.Hit )
+		{
+			var dist = (muzzleTrace.EndPosition - mz.Position).Length;
```
Use `TraceResult.Distance`.

kolexxx commented on code/Player/Player.Flashlight.cs:
```diff
@@ -82,11 +82,58 @@ private void Frame()
+			.Ignore( ActiveCarriable )
+			.Run();
+
+		Vector3 downOffset = Vector3.Down * 2f;
```
Use `var`.

kolexxx commented on code/Player/Player.Flashlight.cs:
```diff
@@ -82,11 +82,58 @@ private void Frame()
+			.Ignore( ActiveCarriable )
+			.Run();
+
+		var distance = (fwdTrace.EndPosition - origin).Length;
```
Use `TraceResult.Distance`.

kolexxx commented on code/Player/Player.Flashlight.cs:
```diff
@@ -82,11 +82,58 @@ private void Frame()
+		if ( distance < pullbackThreshold )
+		{
+			var pullbackAmount = pullbackThreshold - distance;
+			pullbackMult = MathX.Remap( pullbackAmount, 0, pullbackThreshold, 0.0f, 0.045f );
```
`pullbackAmount = pullbackAmount.Remap( 0, pullbackThreshold, 0.0f, 0.045f );`

kolexxx commented on code/Player/Player.Flashlight.cs:
```diff
@@ -82,11 +82,58 @@ private void Frame()
+
+		var distance = (fwdTrace.EndPosition - origin).Length;
+
+		float pullbackMult = 0;
```
Use `var`.
